### ISSUE TITLE
Question : how to addhighlight so that a sentence in 2 row covered by one highlight?

### ISSUE THREAD

**Issue by pluieciel** [pymupdf/PyMuPDF#341]
Hello Team,
I am adding highlights with:
```
text_instances = page.searchFor(text)
for inst in text_instances:
    highlight = page.addHighlightAnnot(inst)
```
but when the sentence is over one line, then, the highlight split into two parts. It looks the same. when you click, it shows two seperate Rects.
Is it possible to add one union highlight over two lines?
Thanks a lot in advance!

**Comment by JorjMcKie** [comment]
I think you mean something like this snippet:
```python
>>> import fitz
>>> doc=fitz.open()
>>> page=doc.newPage()
>>> page.insertText((100, 72), "Line 1")  # insert some text anywhere
1
>>> page.insertText((100, 92) ,"Line 2")  # another text somewhere else
1
>>> rl1=page.searchFor("Line 1")  # search for the two
>>> rl2=page.searchFor("Line 2")  # text pieces
>>> rl1
[Rect(100.0, 60.17499923706055, 129.9639892578125, 75.28900146484375)]
>>> rl2
[Rect(100.0, 80.17500305175781, 129.9639892578125, 95.28900146484375)]
>>> rect = rl1[0] | rl2[0]  # join the rectangles enclosing the text pieces
>>> rect
Rect(100.0, 60.17499923706055, 129.9639892578125, 95.28900146484375)
>>> page.addHighlightAnnot(rect)  # highlight the joined rectangle
'Highlight' annotation on page 0 of <new PDF, doc# 1>
>>> doc.save("highlight.pdf")
>>> 
```
This is how the resulting PDF page looks like:
![grafik](https://user-images.githubusercontent.com/8290722/62661070-ece3b380-b93d-11e9-9a99-b3388c744d2d.png)

Highlight (underline, strikethrough, suqiggle) annotations refer to a **rectangle** only - they know nothing about what the rectangle may be enclosing ... specifically, the rectangle might e.g. also be empty.
It may interest you, that those annotation types also support **quads** (= tilted rectangles). An example is contained in [new-annots.py](https://github.com/pymupdf/PyMuPDF/blob/master/demo/new-annots.py).

**Comment by JorjMcKie** [comment]
Just tried out myself a rather crazy example. I first define 4 arbitrary points on the page and then add a highlight annotation for the quadrilateral defined by those 4 points:
```python
>>> import fitz
>>> doc=fitz.open()
>>> page = doc.newPage()
>>> shape = page.newShape()
>>> p1 = fitz.Point(100,100)
>>> p2 = fitz.Point(300, 150)
>>> p3 = fitz.Point(150,200)
>>> p4 = fitz.Point(300, 400)
>>> shape.drawCircle(p1,2)
Point(98.0, 100.0)
>>> shape.finish()
>>> shape.drawCircle(p2,2)
Point(298.0, 150.0)
>>> shape.finish(color=(1,0,0))
>>> shape.drawCircle(p3,2)
Point(148.0, 200.0)
>>> shape.finish(color=(0,1,0))
>>> shape.drawCircle(p4,2)
Point(298.0, 400.0)
>>> shape.finish(color=(0,0,1))
>>> q = fitz.Quad(p1,p2,p3,p4)
>>> shape.commit()
>>> page.addHighlightAnnot(q)
'Highlight' annotation on page 0 of <new PDF, doc# 1>
>>> doc.save("highlight.pdf")
>>> 
```
The wild result looks like this:
![grafik](https://user-images.githubusercontent.com/8290722/62663073-3800c500-b944-11e9-8a90-a1f1c108118d.png)

**Comment by pluieciel** [comment]
Hi @JorjMcKie 
Thanks a lot for your detailed explaination! I'll try it by myself tomorrow morning.
just to add some details, my need is that one sentence cover right half of line 1 and left half of line 2. and I need one highlight like this:
![image](https://user-images.githubusercontent.com/15624487/62664097-8a5cd880-b97a-11e9-8879-986c320d2e5c.png)
while my previous code gives me two highlights looks the same but not one.

**Comment by JorjMcKie** [comment]
I see. I just made an experiment with a text piece occurring only once on a page, but broken across two lines like your example.
When doing a ``page.searchFor()``, I received a result with **two rectangles**, and I added highlights for both of them:
```python
>>> rl=page.searchFor("depends on how the image was")
>>> rl
[Rect(449.7888488769531, 248.38967895507812, 519.5191040039062, 258.34967041015625), Rect(88.82388305664062, 261.4999084472656, 147.3898162841797, 271.45989990234375)]
>>> page.addHighlightAnnot(rl[0])
'Highlight' annotation on page 0 of v110-changes.pdf
>>> page.addHighlightAnnot(rl[1])
'Highlight' annotation on page 0 of v110-changes.pdf
>>> doc.save("y.pdf")
>>> 
```
... and received this result:
![grafik](https://user-images.githubusercontent.com/8290722/62665261-99c52d00-b94c-11e9-9973-132a65311950.png)

It _**should**_ work in your case, too!

**Comment by JorjMcKie** [comment]
Thank you for your inquiry, anyway.
It has given me reason to consider a way to accomplish this - technically it is possible to create one single textmarker annotation consisting of 2 or more separate, disconnected pieces like in your example.

**Comment by pluieciel** [comment]
My need of this function comes from: when we do highlight manually, it is one part even the sentence covers 2 lines. And I'll need to extract the highlight sentence in some other program.
![image](https://user-images.githubusercontent.com/15624487/62685280-a7b59500-b9c2-11e9-92c1-5127e51407a1.png)
Does this mean a joint of two Rects or it is a strange shage with 8 Xs +8 Ys?

However, with seperate parts of highlight, I will not get one whole sentence when extract from highlighted pdf.
![image](https://user-images.githubusercontent.com/15624487/62685329-bf8d1900-b9c2-11e9-8d6a-9d5863755fe1.png)

So, the 'page.addHighlightAnnot(inst)'.pdf and 'manually HL'.pdf will give different results when I do extraction. 

Thanks a lot for your help ! It would be great if that feature can be added. 
Or I may condiser join the parts back into a whole sentence when extracting (a bit diffcult for me to recognize which ones should be grouped in one sentence since sometimes I get reversed orderd parts).

**Comment by JorjMcKie** [comment]
Yes, I understand.
I have been working to find a solution yesterday, and it seems possible in the following way:
You will need to use ``page.addHighlightAnnot(list)``, where the argument is a list of ``Quad`` objects, which could have been produced by ``page.searchFor(..., quads=True)`` - but not necessarily so.

Let me check how to implement this and decide in which version to publish. Maybe I will add a v1.14.20, or only do this for the coming v1.16.0 (there will not be a version 1.15).

> Does this mean a joint of two Rects or it is a strange shage with 8 Xs +8 Ys?

It is **one** annotation, so has a well defined annotation rectangle. The highlighted parts are defined by a sequence of quads - 1 to many. Each quad (see the docu) consists of 4 ``Point`` objects called UL, UR, LL, LR (upper left, upper right ...) - equivalent to 8 float values. By the books, text marking annotations can contain any number of such quads, and they do not need to be connected to each other at all, or be on consecutive lines.

> Or I may condiser join the parts back into a whole sentence when extracting (a bit diffcult for me to recognize which ones should be grouped in one sentence since sometimes I get reversed orderd parts).

Unexpected text extraction sequences may come from another cause: with PDF you can **_never_** rely on text extraction being in the same sequence as a PDF viewer displays it. In order to reconstruct the original reading sequence, a programmatical effort may be required using bboxes of single words (``page.getTextWords()``), or text spans (as contained in ``page.getText("dict")``), or the coordinates of the quads.

**Comment by pluieciel** [comment]
Thanks a lot!
I'll try your first solution. And looking forward to the coming version!

**Comment by pluieciel** [comment]
Hi @JorjMcKie 
I've tried with `page.addHighlightAnnot(list)`
```
list=[Quad(Point(122.66000366210938, 494.4700012207031), Point(527.02001953125, 494.4700012207031), Point(122.66000366210938, 507.80999755859375), Point(527.02001953125, 507.80999755859375)),
 Quad(Point(96.0, 506.4700012207031), Point(145.72000122070312, 506.4700012207031), Point(96.0, 519.8099975585938), Point(145.72000122070312, 519.8099975585938))]
```
`page.addHighlightAnnot(list)` cannot work
`page.addHighlightAnnot(list[0])` works and highlight the first half of the sentence.

**Comment by JorjMcKie** [comment]
Hi,
as expected: a **_list_** as argument to ``addGighlightAnnot`` is something to come in a later version.
If you add another annot using list[1] however, you should cover the second half.

**Comment by pluieciel** [comment]
Ok, I'll wait for a later version. Since I would need a 'one-block' highlight.
Thanks a lot!

**Comment by JorjMcKie** [comment]
Duplicate of #285

**Comment by pluieciel** [comment]
Hi @JorjMcKie 
Thanks for everything!
I just encountered an error after I update to 14.20:

  File "C:\Users\c30557\AppData\Local\Continuum\anaconda3\lib\site-packages\fitz\fitz.py", line 2015, in __init__
    self._graft_id = TOOLS.gen_id()
NameError: name 'TOOLS' is not defined

Is it due to my environment config or something?
Thanks a lot in advance!

**Comment by JorjMcKie** [comment]
Just installed that same version from PyPI to make sure it works ... and it does!

**Comment by pluieciel** [comment]
I uninstall and install again and it works. not sure why that error happens, sorry.

Many thanks for your always quick help and work of the great tool ! 
It is really powerful and covers all my needs around pdf. :)))

**Comment by JorjMcKie** [comment]
pleased to hear this!
have fun with the tool!
